### PR TITLE
Upgrade node to v18.x

### LINKED ISSUE #1094: Upgrade to Yarn 3 to support Node 16
Try run tests from current master (commit a8dff26dc9485bb338e32b6fe0b532a72309427c) but got error:
```
$ yarn test
➤ YN0000: [@balancer-labs/balancer-js]:
➤ YN0000: [@balancer-labs/balancer-js]:
➤ YN0000: [@balancer-labs/balancer-js]:   sortTokens
➤ YN0000: [@balancer-labs/balancer-js]:     when provided only tokens
➤ YN0000: [@balancer-labs/balancer-js]:       when provided only ERC20s
➤ YN0000: [@balancer-labs/balancer-js]:         √ sorts the tokens in ascending order
➤ YN0000: [@balancer-labs/balancer-js]:       when provided a mix of ERC20s and ETH
➤ YN0000: [@balancer-labs/balancer-js]:         √ sorts ETH as if it were WETH
➤ YN0000: [@balancer-labs/balancer-js]:     when provided additional arrays
➤ YN0000: [@balancer-labs/balancer-js]:       √ sorts the tokens in ascending order
➤ YN0000: [@balancer-labs/balancer-js]:       √ maintains relative ordering with tokens array
➤ YN0000: [@balancer-labs/balancer-js]:
➤ YN0000: [@balancer-labs/balancer-js]:
➤ YN0000: [@balancer-labs/balancer-js]:   4 passing (25ms)
➤ YN0000: [@balancer-labs/balancer-js]:
➤ YN0000: [@balancer-labs/v2-deployments]: Running normal tests...
➤ YN0000: [@balancer-labs/v2-deployments]:
➤ YN0000: [@balancer-labs/v2-deployments]:
➤ YN0000: [@balancer-labs/v2-deployments]:   StablePoolFactory
➤ YN0000: [@balancer-labs/v2-deployments]:     with no previous deploy
➤ YN0000: [@balancer-labs/v2-deployments]:       when forced
➤ YN0000: [@balancer-labs/v2-deployments]:         √ deploys a stable pool factory (1161ms)
➤ YN0000: [@balancer-labs/v2-deployments]:       when not forced
➤ YN0000: [@balancer-labs/v2-deployments]:         √ deploys a stable pool factory (125ms)
➤ YN0000: [@balancer-labs/v2-deployments]:     with a previous deploy
➤ YN0000: [@balancer-labs/v2-deployments]:       when forced
➤ YN0000: [@balancer-labs/v2-deployments]:         √ re-deploys the stable pool factory (274ms)
➤ YN0000: [@balancer-labs/v2-deployments]:       when not forced
➤ YN0000: [@balancer-labs/v2-deployments]:         √ does not re-deploys the stable pool factory (58ms)
➤ YN0000: [@balancer-labs/v2-deployments]:
➤ YN0000: [@balancer-labs/v2-deployments]:   MetaStablePoolFactory
➤ YN0000: [@balancer-labs/v2-deployments]:     with no previous deploy
➤ YN0000: [@balancer-labs/v2-deployments]:       when forced
➤ YN0000: [@balancer-labs/v2-deployments]:         √ deploys a meta stable pool factory (652ms)
➤ YN0000: [@balancer-labs/v2-deployments]:       when not forced
➤ YN0000: [@balancer-labs/v2-deployments]:         √ deploys a meta stable pool factory (361ms)
➤ YN0000: [@balancer-labs/v2-deployments]:     with a previous deploy
➤ YN0000: [@balancer-labs/v2-deployments]:       when forced
➤ YN0000: [@balancer-labs/v2-deployments]:         √ re-deploys the meta stable pool factory (225ms)
➤ YN0000: [@balancer-labs/v2-deployments]:       when not forced
➤ YN0000: [@balancer-labs/v2-deployments]:         √ does not re-deploys the meta stable pool factory (150ms)
➤ YN0000: [@balancer-labs/v2-deployments]:
➤ YN0000: [@balancer-labs/v2-deployments]:
➤ YN0000: [@balancer-labs/v2-deployments]:   8 passing (3s)
➤ YN0000: [@balancer-labs/v2-deployments]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]: Nothing to compile
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:   Aave Asset manager
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:     claimRewards
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       1) "before each" hook: set up asset manager at step "Running shared before each or reverting" for "sends expected amount of stkAave to the rewards contract"
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:   RebalancingRelayer
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:     vault
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       2) "before each" hook: deploy relayer at step "Running shared before each or reverting" for "uses the given vault"
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:   Rewards Asset manager
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:     deployment
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       3) "before each" hook: set up asset manager at step "Running shared before each or reverting" for "different managers can be set for different tokens"
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:   0 passing (581ms)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:   3 failing
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:   1) Aave Asset manager
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:        "before each" hook: set up asset manager at step "Running shared before each or reverting" for "sends expected amount of stkAave to the rewards contract":
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:      HardhatError: HH700: Artifact for contract "TestToken" not found.
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts._handleWrongArtifactForContractName (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:478:11)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts._getArtifactPathFromFiles (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:592:19)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts._getArtifactPath (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:275:17)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts.readArtifact (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:58:26)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Object.deploy (E:\git\balancer\balancer-v2-monorepo\pvt\helpers\src\contract.ts:31:20)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at deployToken (E:\git\balancer\balancer-v2-monorepo\pvt\helpers\src\models\tokens\TokensDeployer.ts:33:18)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at async Promise.all (index 0)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at TokensDeployer.deploy (E:\git\balancer\balancer-v2-monorepo\pvt\helpers\src\models\tokens\TokensDeployer.ts:22:20)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at setup (test\AaveATokenAssetManager.test.ts:24:18)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Context.<anonymous> (test\AaveATokenAssetManager.test.ts:112:27)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Context.<anonymous> (E:\git\balancer\balancer-v2-monorepo\pvt\common\sharedBeforeEach.ts:32:7)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:   2) RebalancingRelayer
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:        "before each" hook: deploy relayer at step "Running shared before each or reverting" for "uses the given vault":
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:      HardhatError: HH700: Artifact for contract "TestToken" not found.
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts._handleWrongArtifactForContractName (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:478:11)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts._getArtifactPathFromFiles (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:592:19)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts._getArtifactPath (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:275:17)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts.readArtifact (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:58:26)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Object.deploy (E:\git\balancer\balancer-v2-monorepo\pvt\helpers\src\contract.ts:31:20)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at TokensDeployer.deployToken (E:\git\balancer\balancer-v2-monorepo\pvt\helpers\src\models\tokens\TokensDeployer.ts:33:18)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Context.<anonymous> (test\RebalancingRelayer.test.ts:30:17)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Context.<anonymous> (E:\git\balancer\balancer-v2-monorepo\pvt\common\sharedBeforeEach.ts:32:7)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:   3) Rewards Asset manager
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:        "before each" hook: set up asset manager at step "Running shared before each or reverting" for "different managers can be set for different tokens":
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:      HardhatError: HH700: Artifact for contract "TestToken" not found.
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts._handleWrongArtifactForContractName (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:478:11)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts._getArtifactPathFromFiles (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:592:19)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts._getArtifactPath (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:275:17)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Artifacts.readArtifact (E:\git\balancer\balancer-v2-monorepo\node_modules\hardhat\src\internal\artifacts.ts:58:26)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Object.deploy (E:\git\balancer\balancer-v2-monorepo\pvt\helpers\src\contract.ts:31:20)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at deployToken (E:\git\balancer\balancer-v2-monorepo\pvt\helpers\src\models\tokens\TokensDeployer.ts:33:18)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at async Promise.all (index 0)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at TokensDeployer.deploy (E:\git\balancer\balancer-v2-monorepo\pvt\helpers\src\models\tokens\TokensDeployer.ts:22:20)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at setup (test\RewardsAssetManager.test.ts:23:18)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Context.<anonymous> (test\RewardsAssetManager.test.ts:79:33)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:       at Context.<anonymous> (E:\git\balancer\balancer-v2-monorepo\pvt\common\sharedBeforeEach.ts:32:7)
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]:

```

Any suggestions how to fix this?

Win10, mingw64, node 16.13, npm 8.1.4, yarn 2.4.1. Same error when run tests under WSL.

Building output in win:
```
$ yarn build
➤ YN0000: [@balancer-labs/v2-deployments]: ../../pvt/common/setupTests.ts(7,32): error TS2307: Cannot find module '@balancer-labs/balancer-js' or its corresponding type declarations.
➤ YN0000: [@balancer-labs/balancer-js]:
➤ YN0000: [@balancer-labs/balancer-js]: src/index.ts → dist/index.umd.js, dist/index.js, dist/index.esm.js...
➤ YN0000: [@balancer-labs/balancer-js]: (!) Missing global variable names
➤ YN0000: [@balancer-labs/balancer-js]: Use output.globals to specify browser global variable names corresponding to external modules
➤ YN0000: [@balancer-labs/balancer-js]: @ethersproject/abi (guessing 'abi')
➤ YN0000: [@balancer-labs/balancer-js]: @ethersproject/constants (guessing 'constants')
➤ YN0000: [@balancer-labs/balancer-js]: @ethersproject/bignumber (guessing 'bignumber')
➤ YN0000: [@balancer-labs/balancer-js]: @ethersproject/bytes (guessing 'bytes')
➤ YN0000: [@balancer-labs/balancer-js]: @ethersproject/abstract-signer (guessing 'abstractSigner')
➤ YN0000: [@balancer-labs/balancer-js]: @ethersproject/address (guessing 'address')
➤ YN0000: [@balancer-labs/balancer-js]: created dist/index.umd.js, dist/index.js, dist/index.esm.js in 1.9s
➤ YN0000: [@balancer-labs/balancer-js]:
➤ YN0000: [@balancer-labs/balancer-js]: src/index.ts → dist/index.d.ts...
➤ YN0000: [@balancer-labs/balancer-js]: created dist/index.d.ts in 1.1s
➤ YN0000: [@balancer-labs/v2-asset-manager-utils]: Nothing to compile
```

### PR DESCRIPTION
**Pull request by jubeira**
# Description

Upgrade engine to Node 18 and package manager to Yarn V4.

Tested locally with Node 18.0.0, 18.12.1 and 18.15.0 (LTS).

The current default version used by the CI can be found [here](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#nodejs). As of end of April-23, `ubuntu-latest` is 22, and Node 18 is `18.16.0`. This version happens to have a bug that makes hardhat skip compiling when downloading solc from time to time (see https://github.com/NomicFoundation/hardhat/issues/3877). Therefore, the minor version also has to be specified in the docs and in the actions yaml file to prevent using it.

Some notes:
- I've noticed that building `deployments` as a standalone package fails. In master, building all the workspaces still works, but it fails the first time when upgrading to Node 18. `--topological` flag was added to the build step to ensure that `balancer-js` is built before `v2-deployments`. In short, it forces yarn to builld dependencies before building packages (just applies to dependencies, not dev dependencies - [ref](https://yarnpkg.com/cli/workspaces/foreach#options)). EDIT: flag removed in https://github.com/balancer/balancer-v2-monorepo/pull/2020/commits/8b5acf3d8e5480b37e8a494cdad61991a1446691; the race condition is still present. Added a note in the main readme to make a difference in the first build to avoid confusion on this topic.

- **[edit: solved in Yarn V4]** There's an issue with the checksum of the downloaded packages in CI that breaks the build, which doesn't happen when building locally using Yarn 3. It can be solved adding `checksumBehavior: "update"` to `.yarnrc.yml` to circumvent it. See refs:
  - https://github.com/yarnpkg/berry/issues/1142.
   - https://github.com/pastelsky/bundlephobia/issues/702#issuecomment-1299431888.

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [x] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [ ] The diff is legible and has no extraneous changes (see comments)
- [ ] Complex code has been commented, including external interfaces --> N/A
- [ ] Tests are included for all code paths --> N/A
- [x] The base branch is either `master`, or there's a description of how to merge

## Issue Resolution

Closes #1094.

### CODE REVIEW

**Comment by jubeira** [comment]
Not sure why CI is not running, but it should be green:  https://github.com/jubeira/balancer-v2-monorepo/pull/8.

File: .yarnrc.yml [hunk 1]
```diff
@@ -9,4 +10,5 @@ plugins:
   - path: .yarn/plugins/@yarnpkg/plugin-version.cjs
     spec: "@yarnpkg/plugin-version"
 
-yarnPath: .yarn/releases/yarn-berry.cjs
+yarnPath: .yarn/releases/yarn-3.3.0.cjs
+checksumBehavior: "update"
```
Comment (jubeira) [inline .yarnrc.yml:14]: I can confirm that this can be removed in Yarn 4: https://github.com/jubeira/balancer-v2-monorepo/pull/9.

I didn't find any other sensible way of dealing with this so far, and I'm not sure it's worth investing more time in it.
I'd either go with this flag, or move to Yarn 4.

Comment (nventuro) [reply .yarnrc.yml:14]: Is there a reason to use 3 over 4?

File: pkg/deployments/package.json [hunk 2]
```diff
@@ -31,8 +31,10 @@
     "test": "hardhat test ./**/test/*.ts",
     "ci:prepare-config": "ts-node ci/prepare-config.ts"
   },
+  "dependencies": {
+    "@balancer-labs/balancer-js": "workspace:*"
+  },
```
Comment (nventuro) [inline pkg/deployments/package.json:36]: Why this change? The deployments package does not depend on balancer-js

File: pkg/solidity-utils/test/CodeDeployer.test.ts [hunk 3]
```diff
@@ -35,7 +35,7 @@ describe('CodeDeployer', function () {
       const receipt = await (await factory.deploy(data)).wait();
       const event = expectEvent.inReceipt(receipt, 'CodeDeployed');
 
-      expect(await ethers.provider.getCode(event.args.at)).to.equal(data);
+      expect(await ethers.provider.getCode(event.args[0])).to.equal(data);
```
Comment (nventuro) [inline pkg/solidity-utils/test/CodeDeployer.test.ts:38]: What changed here?

Comment (jubeira) [reply pkg/solidity-utils/test/CodeDeployer.test.ts:38]: When I run this in node 18 `event.args` contains an array of strings, so `event.args.at` is not the actual address in the event.

Not sure why this happens; this was the only test that had this issue. I've noticed that the address in this event is not indexed; can't confirm if it's related with that.

File: pkg/deployments/package.json (hunk 2 above)
Comment (jubeira) [reply pkg/deployments/package.json:36]: there are some deployment tasks that import a few types from `balancer-js`.

I only got the error in the first build and not always, saying it didn't find `@balancer-js`. So it seems to be a race condition.

I can try to reproduce it again to get the exact message if needed.

File: .yarnrc.yml (hunk 1 above)
Comment (jubeira) [reply .yarnrc.yml:14]: Yarn 4 latest version is `4.0.0-rc29`, so technically it's still not released yet. 

But other than that both seem to work. Besides the checksum fix, apparently in version 4 they've bundled the plugins we use, so there's no need to declare them in the `.yarnrc.yml` file.

File: pkg/deployments/package.json (hunk 2 above)
Comment (jubeira) [reply pkg/deployments/package.json:36]: This is the error message:

```
➤ YN0000: [@balancer-labs/v2-deployments]: ../../pvt/common/setupTests.ts(7,32): error TS2307: Cannot find module '@balancer-labs/balancer-js' or its corresponding type declarations.
```

Shouldn't `pvt/common` depend on `balancer-js` as well? (note that adding this dependency doesn't fix the issue by itself).

File: pkg/solidity-utils/test/CodeDeployer.test.ts (hunk 3 above)
Comment (jubeira) [reply pkg/solidity-utils/test/CodeDeployer.test.ts]: Update: perhaps this is a bug in ethers?

The problem was the oarameter name. `at` is a method for `Array`, and apparently it collided with the parameter name of the event. Then, the contents of `event.args` were different than the usual ones.

Replacing `at` with `destination` fixed it.

File: pkg/deployments/package.json (hunk 2 above)
Comment (jubeira) [reply pkg/deployments/package.json:36]: One more note about this one:
- I understand that compilation fails without building `balancer-js` because of this dependency here: https://github.com/balancer-labs/balancer-v2-monorepo/blob/116f33b9000aa7e18b35aa28f1e32cae35eb9039/pkg/deployments/hardhat.config.ts#L5 which I think can be removed.
- Even if we can 'compile' `deployments` without `balancer-js`, fork tests won't run because there are some type imports that need it. Since the fork tests are outside `src`, that dependency will go unnoticed when running `build`. Therefore, I think it does make sense to add `balancer-js` as a dependency.
- (related) I've noticed that `v2-common` and `v2-helpers` also use `balancer-js`, but they don't declare it as a dependency. Those two packages also don't have a build task, but if you tried building them with `tsc` (just like `deployments` build task) they'd fail. So maybe they should depend on `balancer-js` as well?

File: .yarnrc.yml (hunk 1 above)
Comment (jubeira) [reply .yarnrc.yml:14]: Update: the current verson is rc31 and it also works.

According to https://github.com/yarnpkg/berry/discussions/4895#discussioncomment-3726438, it should be fine to use an RC version. I don't think they'll actually break anything we use when they release 4.0.0.

**Comment by TomAFrench** [comment]
Merged back in master as refreshing yarn.lock should hopefully help with the cache issue.

File: pkg/deployments/package.json (hunk 2 above)
Comment (nventuro) [reply pkg/deployments/package.json:38]: > fork tests won't run because there are some type imports that need it. (...) Therefore, I think it does make sense to add balancer-js as a dependency.

This is not what 'dependency' means in the context of npm packages, however. A 'dependency' package is one that must be installed in order to run (use) the package in question, and will therefore be automatically installed when installing the root package.

`deployments` has no dependencies: you don't need any other package in order to use it (to load the json files, etc.). If we do need a package in order to e.g. run tests, then it's a `devDependency`, which means it must be installed to do development (same as linters, test tools, compilers, etc) - but it's not a dependency of the package _as an end product_.

Comment (nventuro) [reply pkg/deployments/package.json:38]: (we don't want to force users to install `balancer-js` when installing `deployments`).

Comment (jubeira) [reply pkg/deployments/package.json:38]: In that case, shouldn't we remove the `tsc` step from the `build` script? Otherwise `yarn build` will just fail

**Comment by jubeira** [comment]
@nventuro build seems to work in CI repeatedly https://github.com/balancer/balancer-v2-monorepo/actions/runs/4376881980/jobs/7659617952, but the is still a race condition when building everything the first time (`deployments` expects `balancer-js` to be finished). In the CI logs, you can see that `balancer-js` ends before `deployments` is built, so it doesn't cause issues.

I can reproduce the issue locally with Node 18, and I believe this also happens with Node 14 but I can't reproduce it locally.

File: .yarnrc.yml (hunk 1 above)
Comment (jubeira) [reply .yarnrc.yml]: Updated to rc42 in https://github.com/balancer/balancer-v2-monorepo/pull/2020/commits/c4b55babbc476a96a39f8c1f77d0964f5a987b13, and it still works.

File: pkg/deployments/package.json (hunk 2 above)
Comment (jubeira) [reply pkg/deployments/package.json]: I've added a note referring to this, and removed the `--topological` flag.

Forcing to build the packages in sequence and adding the explicit dependencies is an overkill; the note should be enough to consider in first-time builds. In CI this doesn't seem to be an issue.

File: README.md [hunk 4]
```diff
@@ -31,7 +31,7 @@ Active development occurs in this repository, which means some contracts in it m
 
 ## Pre-requisites
 
-The build & test instructions below should work out of the box with Node ^14.18.0. (Please note that it needs Node 14 specifically, and will NOT work with Node 16 or higher. Minor version should be at least 18).
+The build & test instructions below should work out of the box with Node 18. More specifically, it is recommended to use the LTS version ^18.15; Node 19 is not supported.
```
Comment (EndymionJkb) [inline README.md:34]:
```suggestion
The build & test instructions below should work out of the box with Node 18. More specifically, it is recommended to use the LTS version 18.15; Node 19 and higher are not supported.
```
We need 18.15 exactly, right? And you mentioned 20's out :) TMI to say why we need 15 specifically?

File: README.md [hunk 5]
```diff
@@ -52,8 +52,16 @@ $ git clone --recurse-submodules https://github.com/balancer-labs/balancer-v2-mo
 
 Before any tests can be run, the repository needs to be prepared:
 
+### First time build
+
 ```bash
 $ yarn # install all dependencies
+$ yarn workspace @balancer-labs/balancer-js buildbuild # build balancer-js first
```
Comment (EndymionJkb) [inline README.md:59]:
```suggestion
$ yarn workspace @balancer-labs/balancer-js build # build balancer-js first
```
Duplicated?

**Review by EndymionJkb — COMMENTED** [review]
What are the CodeDeployerFactory and common/package.json changes? Part of another PR?

File: README.md (hunk 4 above)
Comment (jubeira) [reply README.md]: How about now?

**Comment by jubeira** [comment]
> What are the CodeDeployerFactory and common/package.json changes? Part of another PR?

This is part of this PR. With the upgrade, `at` is a reserved member when dealing with events in `ethers`. To prevent the collision, I've renamed that particular appearance.

**Review by EndymionJkb — APPROVED** [review]
I think we need to not wait until right before EOL to update things: it *might* be less painful. Though some of these issues were quite recent, so I'm not even sure that's true.

Anyway, kudos to Juani for tracking down and resolving the myriad third party bugs and other hurdles that dragged this out so long.
